packages/cli/src/commands/devHandler.js

```diff
@@ -23,6 +24,71 @@ export const handler = async ({
 }) => {
   const rwjsPaths = getPaths()
 
+  // Starting values of ports from config (redwood.toml)
+  let apiPreferredPort = parseInt(getConfig().api.port)
+  let webPreferredPort = parseInt(getConfig().web.port)
+
+  // Assume we can have the ports we want
+  let apiAvailablePort = apiPreferredPort
+  let apiPortChangeNeeded = false
+  let webAvailablePort = webPreferredPort
+  let webPortChangeNeeded = false
+
+  // Check api port
+  if (side.includes('api')) {
+    apiAvailablePort = await getFreePort(apiPreferredPort)
+    if (apiAvailablePort === -1) {
+      errorTelemetry(
+        process.argv,
+        `Error could not determine a free port for the api server`
+      )
+      console.error(`Error could not determine a free port for the api server`)
+      process.exit(-1)
+    }
+    apiPortChangeNeeded = apiAvailablePort !== apiPreferredPort
+  }
+
+  // Check web port
+  if (side.includes('web')) {
+    // Extract any ports the user forwarded to the webpack server and prefer that instead
+    const forwardedPortMatches = [
+      ...forward.matchAll(/\-\-port(\=|\s)(?<port>[^\s]*)/g),
+    ]
+    if (forwardedPortMatches.length) {
+      webPreferredPort = forwardedPortMatches.pop().groups.port
+    }
+
+    webAvailablePort = await getFreePort(webPreferredPort, [
+      apiPreferredPort,
+      apiAvailablePort,
+    ])
+    if (webAvailablePort === -1) {
+      errorTelemetry(
+        process.argv,
+        `Error could not determine a free port for the web server`
+      )
+      console.error(`Error could not determine a free port for the web server`)
+      process.exit(-1)
+    }
+    webPortChangeNeeded = webAvailablePort !== webPreferredPort
+  }
+
+  // Check for port conflict and exit with message if found
+  if (apiPortChangeNeeded || webPortChangeNeeded) {
+    let message = `The currently configured ports for the development server are unavailable. Suggested changes to your ports, which can be changed in redwood.toml, are:\n`
+    message += apiPortChangeNeeded
+      ? `  - API to use port ${apiAvailablePort} instead of your currently configured ${apiPreferredPort}\n`
+      : ``
+    message += webPortChangeNeeded
+      ? `  - Web to use port ${webAvailablePort} instead of your currently configured ${webPreferredPort}\n`
+      : ``
+    console.error(message)
+    console.error(
+      `Cannot run the development server until your configured ports are changed or become available.`
+    )
+    process.exit(-1)
```

Comment:
I don't think you ever exit with a negative number; the most I know though is 0 is for success and 1-`Infinity` is error:

```suggestion
    process.exit(1)
```

Reply:
Okay I'll keep them positive from now on, thanks.